File: leetcode/11. Container With Most Water.ts
//https://leetcode.com/problems/container-with-most-water
function maxArea(height: number[]): number {

    let max = 0
    let left = 0
    let right = height.length-1

    while(left < right) {
        const h = Math.min(height[left], height[right])
        const d = right - left
        const area = h*d

        if(max < area) max = area

        if(height[left] <= height[right]) {
            left++
        } else {
            right--
        }
    }

    return max
};

File: leetcode/11. Container with most water.go
func maxArea(height []int) int {
	left := 0
	right := len(height) - 1
	res := 0

	for left < right {
		area := getArea(height, left, right)
		if area > res {
			res = area
		}

		if height[left] > height[right] {
			right--
		} else {
			left++
		}
	}

	return res
}

func getArea(height []int, left, right int) int {
	return min(height[left], height[right]) * (right - left)
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}
